src/MiniCover/Reports/Clover/CloverReport.cs

```diff
@@ -0,0 +1,198 @@
+                new XAttribute(XName.Get("methods"), counter.Methods),
+                new XAttribute(XName.Get("coveredmethods"), counter.CoveredMethods),
+                new XAttribute(XName.Get("elements"), counter.Elements),
+                new XAttribute(XName.Get("coveredelements"), counter.CoveredElements)
+            );
+
+            if(counter.Lines > 0)
+            {
+                element.Add(new XAttribute(XName.Get("loc"), counter.Lines));
+                element.Add(new XAttribute(XName.Get("ncloc"), counter.Lines));
+            }
+            if(counter.Classes > 0)
+            {
+                element.Add(new XAttribute(XName.Get("classes"), counter.Classes));
+            }
+            if(counter.Files > 0)
+            {
+                element.Add(new XAttribute(XName.Get("files"), counter.Files));
+            }
+            if(counter.Packages > 0)
+            {
+                element.Add(new XAttribute(XName.Get("packages"), counter.Packages));
+            }
+
+            return element;
+        }
+
+        private static CloverCounter CountProjectMetrics(InstrumentationResult result, IDictionary<int, int> hits)
+        {
+            return result.Assemblies
+                .Select(t => CountPackageMetrics(t, hits))
+                .Aggregate((prev, next) =>
+                {
+                    prev.Add(next);
+                    prev.Packages += 1;
+                    return prev;
+                });;
+        }
+
+        private static CloverCounter CountPackageMetrics(InstrumentedAssembly assembly, IDictionary<int, int> hits)
+        {
+            return assembly.SourceFiles
+                .Select(t => CountFileMetrics(t.Value, hits))
+                .Aggregate((prev, next) =>
+                {
+                    prev.Add(next);
+                    prev.Files += 1;
+                    return prev;
+                });
+        }
+
+        private static CloverCounter CountFileMetrics(SourceFile file, IDictionary<int, int> hits)
+        {
+            var counter = CountMetrics(file.Instructions, hits);
+            counter.Lines = file.Instructions.Max(instruction => instruction.EndLine);
+            counter.Classes = file.Instructions.GroupBy(t => t.Class).Count();
+            return counter;
+        }
+
+        private static CloverCounter CountMetrics(IEnumerable<InstrumentedInstruction> instructions, IDictionary<int, int> hits)
+        {
+            return new CloverCounter
+            {
+                Statements = instructions.Count(),
+                CoveredStatements = instructions
+                    .Where(instruction => hits.ContainsKey(instruction.Id))
+                    .Count(),
+                Methods = instructions
+                    .GroupBy(instruction => instruction.MethodFullName)
+                    .Count(),
+                CoveredMethods = instructions
+                    .Where(instruction => hits.ContainsKey(instruction.Id))
```

Comment:
it looks like multiple enumerations. maybe this 'where' can we do before

Reply:
ok